opensafely/check.py

```diff
@@ -0,0 +1,115 @@
+from opensafely._vendor import requests
+from opensafely._vendor.ruamel.yaml import YAML
+import os
+import re
+import glob
+import subprocess
+
+DESCRIPTION = "Check the opensafely project for correctness"
+
+RESTRICTED_DATASETS = {"icnarc": ["admitted_to_icu"]}
+
+# todo: dummy location and data
+PERMISSIONS_URL = "https://raw.githubusercontent.com/Jongmassey/research_repository_permissions/main/repository_permissions.yaml"
+
+
+def add_arguments(parser):
+    pass
+
+
+def main():
+    repo_name = get_repository_name()
+    permissions = get_datasource_permissions(PERMISSIONS_URL)
+    allowed_datasets = get_allowed_datasets(repo_name, permissions)
+    datasets_to_check = {
+        k: v
+        for k, v in RESTRICTED_DATASETS.items()
+        if k not in allowed_datasets
+    }
+    files_to_check = get_local_py_files()
+
+    found_datasets = {
+        dataset: check_dataset(functions, files_to_check)
+        for dataset, functions in datasets_to_check.items()
+        if check_dataset(functions, files_to_check)
+    }
+
+    if found_datasets:
+        exit(print_violations(found_datasets))
+    else:
+        print("Success")
+
+
+def print_violations(found_datasets):
+    s = "Usage of restricted datasets found:\n"
+    for d, functions in found_datasets.items():
+        s = s + (f"{d}:\n")
+        for fn, files in functions.items():
+            s = s + (f"- {fn}\n")
+            for f, lines in files.items():
+                s = s + (f"  - {f}:\n")
+                for ln, line in lines.items():
+                    s = s + (f"    line {ln}: {line}\n")
```

Comment:
nit: strings are immutable in python, so string concatenation is generally avoided as it allocation heavy.

The typical idiom is to join a list of lines, e.g.

```python
lines = []
...
lines.append(line)
...
return "\n".join(lines)
```

Which also saves you have having to append "\n" to every line's text.

However, for more fun, you can stream it!

```python
def print_violations(found_datasets):
    yield "Usage of restricted datasets found:"
    for d, functions in found_datasets.items():
        yield f"{d}:"
        for fn, files in functions.items():
            yield f"- {fn}"
            for f, lines in files.items():
                yield f"  - {f}:"
                for ln, line in lines.items():
                    yield f"    line {ln}: {line}"

```

And use it like so:

```python
voilations = "\n".join(print_violations(found_violations))
sys.exit(violations)
```

Reply:
Good to know a concise way of streaming it in Python. I miss my StringBuider classes!